controller/service.go

```diff
@@ -144,39 +166,46 @@ func (c *controller) clearServiceState(key string, svc *v1.Service) {
 	svc.Status.LoadBalancer = v1.LoadBalancerStatus{}
 }
 
-func (c *controller) allocateIP(key string, svc *v1.Service) (net.IP, error) {
-	clusterIP := net.ParseIP(svc.Spec.ClusterIP)
-	if clusterIP == nil {
+func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error) {
+	var serviceIPFamily ipfamily.IPFamily
+	var err error
+	if svc.Spec.ClusterIPs != nil {
+		serviceIPFamily, err = ipfamily.IPFamilyForAddresses(svc.Spec.ClusterIPs)
+	} else {
 		// (we should never get here because the caller ensured that Spec.ClusterIP != nil)
 		return nil, fmt.Errorf("invalid ClusterIP [%s], can't determine family", svc.Spec.ClusterIP)
 	}
-	isIPv6 := clusterIP.To4() == nil
+
+	if err != nil {
```

Comment:
We can avoid the extra nesting by checking for nill and returning (as the original code was doing).
if svc.Spec.ClusterIPs != nil {
	return nil, fmt.Errorf("invalid ClusterIP [%s], can't determine family", svc.Spec.ClusterIP)
}

Reply:
u meant == nil ok removed else

Reply:
Yep sorry